Fix multiple choice answer logic and update tests to check that

pjanik opened the pull request:
[#158162480]

The previous code was assuming that a question without a correct answer will always return `isCorrect=null` properties for every choice. That's not true. Sometimes all the choices have `isCorrect=false` what also means that there's no correct answer. New code decides whether given question has the correct answer or not by looking at all the choices and checking if there's at least one with `isCorrect=true`.

@ekosmin, could you take a look please?
You can also be interested in tests, as this case is pretty similar to your button case. I have two separate components. Instead of using `render` or `mount`, I render shallowly both of them and check whether the top one is passing right properties to its children. Then I test child component separately. Not saying that's it's better or wrose. Just more unit oriented. Other thing is that we definitely have great test coverage in this project, but I'm curious how painful would be some major components refactoring and/or shuffling. :wink: Although I don't see a way to make it simpler, perhaps just set of new tests and removal of the old ones. We could focus testing on the rendered content (like text), not React components, but it seems like more work and won't let you test everything you want that easily.

ekosmin commented on test/components/dashboard/multiple-choice-answer_spec.js:
```diff
@@ -48,4 +59,75 @@ describe('<MultipleChoiceAnswer />', () => {
   })
+
+  describe('when there is no choice with isCorrect property equal to `true`', () => {
+    it('should pass `correctAnswerDefined=false` prop to choices', () => {
```
It seems like the logic worth testing here is that `correctAnswerDefined` is computed properly, more so than that it's passed correctly as a prop. I'd consider just testing that function on its own - that test might be less likely to break in the future.

pjanik replied:
Yeap. But on the other hand, you test less funcionality. I'm usually trying to cover more code and functionality as long as it's easy.

pjanik commented on js/components/dashboard/multiple-choice-answer.js:
```diff
@@ -67,17 +67,20 @@ export class Choice extends PureComponent {
+  get correctAnswerDefined () {
+    const { question } = this.props
+    const choices = question.get('choices')
+    return !!choices.find(c => c.get('isCorrect'))
```
Thanks, I didn't know it, and I felt I'm missing some function. ;)

pjanik commented on js/components/dashboard/multiple-choice-answer.js:
```diff
@@ -24,11 +24,11 @@ export class Choice extends PureComponent {
+    const { choice, selected, correctAnswerDefined } = this.props
     const isCorrect = choice.get('isCorrect')
-    if (isCorrect === true) {
+    if (correctAnswerDefined && isCorrect) {
```
Yeap, they seem similar, but they are all different actually.

pjanik commented on test/components/dashboard/multiple-choice-answer_spec.js:
```diff
@@ -48,4 +59,75 @@ describe('<MultipleChoiceAnswer />', () => {
+  describe('when there is at least one choice with isCorrect property equal to `true`', () => {
+    it('should pass `correctAnswerDefined=true` prop to choices', () => {
+      const answer = fromJS({ answer: [ { id: 1, choice: 'choice_1' } ] })
+      const question = fromJS({ choices: [ { id: 1, choice: 'choice_1', isCorrect: true }, { id: 2, choice: 'choice_2', isCorrect: false } ] })
```
It's a matter of taste, although I prefer that and keeping tests simple. That way you look at one test and know what's going on, without having to guess the whole logic in given test suite. I do it only when really necessary.

pjanik commented on test/components/dashboard/multiple-choice-answer_spec.js:
```diff
@@ -27,12 +27,23 @@ describe('<MultipleChoiceAnswer />', () => {
+      it('should render empty checkmark icon only', () => {
+        const answer = fromJS({ answer: [ { id: 2, choice: 'choice_2' } ], isCorrect: null })
+        const question = fromJS({ choices: [ { id: 1, choice: 'choice_1', isCorrect: null }, { id: 2, choice: 'choice_2', isCorrect: null } ] })
+        const wrapper = render(<MultipleChoiceAnswer showFullAnswer={false} question={question} answer={answer} />)
```
Yes, good catch, thanks.

pjanik commented:
Thanks Eli. I've addressed most of the comments.

> I also very much agree that our tests are great now, but that they could cause trouble moving forward. Even moving the student sort drop-down to the right of the screen broke tests since I moved it between components :(

I think that's to be expected and it's normal. Even in the back-end world. If you move a method between controllers, or even function between modules, you will break tests and will have to update them. I think it's fine. And even helpful. You know which part of the code your change affects. The only thing is that the front-end code tends to be moved around more often, as it's quite often required by design update. I guess back-end or libraries are more stable. That's what put me off from React testing before, but I think it's not right excuse. :wink:

>  I really like the idea of testing, but to be honest I'm not convinced it's saved me more than it's cost me over my life.

Yes, if you think about small projects that you've worked on yourself, and you consider **only development time**, the return of the investment is perhaps negative. :wink: I guess it helps later, when project grows, or when other developer works on your code. You have confidence that things really work. The tests you break and have to fix are a good sign that refactoring is complete (when the coverage is good enough). You're more likely to add quality code. And so on, and so forth.
